### PR TITLE
pipe_tasks tickets/DM-2797: HSC improvements to colorterms

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: python/lsst/pipe/tasks/colorterms.py [hunk 1]
```diff
@@ -19,130 +19,157 @@
...
+    c2 = Field(dtype=float, default=0.0, doc="Second-order parameter")
+
+    # the following is desired to allow positional data, but is impossible due to DM-2381
+    # def __init__(self, primary, secondary, c0=0.0, c1=0.0, c2=0.0):
+    #     """!Construct a Colorterm
+
+    #     This overrides the default constructor for improved error detection
+    #     and to allow positional arguments
+    #     """
+    #     Config.__init__(self, primary=primary, secondary=secondary, c0=c0, c1=c1, c2=c2)
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: Now that you have a workaround form DM-2831, can you re-enable this?  If not, probably best to just delete it.

File: python/lsst/pipe/tasks/colorterms.py [hunk 2]
```diff
@@ -19,130 +19,157 @@
...
+
+    def transformSource(self, source):
+        """!Transform the brightness of a source
+
+        @param[in] source  source whose brightness is to be converted; must support get(filterName)
+                    (e.g. source.get("r")) method, as do afw::table::Source and dicts.
+        @return the transformed source magnitude
         """
-        return Colorterm._activeColorterms.get(band) if Colorterm._activeColorterms else None
+        return self.transformMags(source.get(self.primary), source.get(self.secondary))
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: Need to state what unit source "brightness" is in - magnitudes, I assume, not fluxes?

File: python/lsst/pipe/tasks/colorterms.py [hunk 3]
```diff
@@ -19,130 +19,157 @@
...
+    #     """!Construct a ColortermDictConfig
 
-        if not colorterms:
-            return primary
+    #     Overrides the default constructor for improved safety
+    #     and to allow the data to be a positional argument
+    #     """
+    #     if dict is None: # mutable objects should not be used as defaults
+    #         dict = {}
+    #     Config.__init__(self, dict=dict)
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: Again, enable with workaround or remove.

File: python/lsst/pipe/tasks/colorterms.py [hunk 4]
```diff
@@ -19,130 +19,157 @@
...
-        ct = colorterms[band]
-        
-        return Colorterm.transformMags(band, source.get(ct.primary), source.get(ct.secondary),
-                                       reverse, colorterms)
 
-    @staticmethod
-    def transformMags(band, primary, secondary, reverse=False, colorterms=None):
-        """!Transform the magnitudes *primary* and *secondary* to the specified *band* and return it.
+class ColortermDictConfig(Config):
+    """!A config containing dict: a dict of filterName: Colorterm
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: While it's more explicit, I actually find that mentioning the name of the config field here makes the documentation a bit harder to read, especially with two colons that work differently.  I'd just say "A config containing a mapping from filterName to Colorterm".  Same with `ColortermLibraryConfig`.

File: python/lsst/pipe/tasks/colorterms.py [hunk 5]
```diff
@@ -19,130 +19,157 @@
...
+    pass
 
-                Colorterm._activeColorterms = None
 
-    @staticmethod
-    def getColorterm(band):
-        """!Return the Colorterm for the specified band (or None if unknown)
+class Colorterm(Config):
+    """!Configuration describing a Colorterm
+
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: Trailing whitespace on this line.

File: python/lsst/pipe/tasks/colorterms.py [hunk 6]
```diff
@@ -19,130 +19,157 @@
...
+    #     """!Construct a Colorterm
+
+    #     This overrides the default constructor for improved error detection
+    #     and to allow positional arguments
+    #     """
+    #     Config.__init__(self, primary=primary, secondary=secondary, c0=c0, c1=c1, c2=c2)
+
+    def transformSource(self, source):
+        """!Transform the brightness of a source
+
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: Trailing whitespace on this line.

File: python/lsst/pipe/tasks/colorterms.py [hunk 7]
```diff
@@ -19,130 +19,157 @@
...
+        We use the group of color terms in the library that matches the refCatName.
+        If the refCatName exactly matches an entry in the library, that
+        group is used; otherwise if the refCatName matches a single glob (shell syntax,
+        e.g., "sdss-*" will match "sdss-dr8"), then that is used.  If there is no
+        exact match and no unique match to the globs, we raise an exception.
+
+        @param filterName  name of filter
+        @param refCatName  reference catalog name or glob expression; if a glob expression then
+            there must be exactly one match in the library
+        @param[in] doRaise  if True then raise ColortermNotFoundError if no suitable Colorterm found;
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: I don't see any indication in the implementation that this particular exception is raised.

Comment (TallJimbo) [inline python/lsst/pipe/tasks/colorterms.py]: To clarify, I mean that other exception types could also be raised.

File: python/lsst/pipe/tasks/photoCal.py [hunk 8]
```diff
@@ -401,7 +430,7 @@ def extractMagArrays(self, matches, filterName, sourceKeys):
             # this is an unpleasant hack; see DM-2308 requesting a better solution
             self.log.warn("Source catalog does not have flux uncertainties; using sqrt(flux).")
             srcFluxErrArr = np.sqrt(srcFluxArr)
-
+
```
Comment (TallJimbo) [inline python/lsst/pipe/tasks/photoCal.py]: Trailing whitespace (and an unnecessary change).

File: python/lsst/pipe/tasks/colorterms.py (hunk 1 above)
Comment (r-owen) [inline python/lsst/pipe/tasks/colorterms.py]: I tried and it still fails. I'll remove the commented-out code but not squash that change so I can get it back from git.

File: python/lsst/pipe/tasks/colorterms.py (hunk 2 above)
Comment (r-owen) [inline python/lsst/pipe/tasks/colorterms.py]: The word "magnitude" was used everywhere in the doc string, so I thought it was clear, but I'll use "brightness" and "(magnitude)" to see if that helps.

File: python/lsst/pipe/tasks/colorterms.py (hunk 7 above)
Comment (r-owen) [inline python/lsst/pipe/tasks/colorterms.py]: I fixed it so expected errors are all ColortermNotFoundError and raising them is controlled by doRaise. Unexpected errors will break through.

File: python/lsst/pipe/tasks/colorterms.py (hunk 4 above)
Comment (r-owen) [inline python/lsst/pipe/tasks/colorterms.py]: I simplified the description.

**Comment by jdswinbank** [comment]
This work is long since complete.
